File: ui/src/mocks/picks.ts
import type {
  MatchRow,
  PickResponse,
  PicksListResponse,
  PicksSummary,
  PickView,
  RivalPick,
  UserPick,
} from "../lib/contract";
import { PICK_SCHEMA_VERSION } from "../lib/contract";
import rivalsFixture from "./rival-picks.json";

export const MOCK_RIVALS = rivalsFixture as RivalPick[];

const STORE_KEY = "golavo-picks-v1";

export class MockPickError extends Error {
  constructor(
    readonly status: number,
    readonly reasonCode: string,
    message: string,
  ) {
    super(message);
  }
}

function readStore(): Record<string, UserPick> {
  try {
    const value = JSON.parse(localStorage.getItem(STORE_KEY) ?? "{}") as unknown;
    return value && typeof value === "object" ? (value as Record<string, UserPick>) : {};
  } catch {
    return {};
  }
}

function writeStore(store: Record<string, UserPick>): void {
  localStorage.setItem(STORE_KEY, JSON.stringify(store));
}

function ordered(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(ordered);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([key, item]) => [key, ordered(item)]),
    );
  }
  return value;
}

async function sha256(value: unknown): Promise<string> {
  const bytes = new TextEncoder().encode(JSON.stringify(ordered(value)));
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

async function matches(): Promise<MatchRow[]> {
  const mod = await import("./matches.json");
  return (mod.default as { matches: MatchRow[] }).matches;
}

function outcome(home: number, away: number): "home" | "draw" | "away" {
  return home > away ? "home" : home < away ? "away" : "draw";
}

async function lock(record: UserPick): Promise<UserPick> {
  if (record.status === "locked") return record;
  const locked: UserPick = {
    ...record,
    status: "locked",
    lock_at_utc: record.lock_at_utc,
    locked_at_utc: record.lock_at_utc,
    updated_at_utc: record.lock_at_utc,
  };
  const stable = { ...locked, pick_id: undefined, payload_sha256: undefined };
  const idHash = await sha256(stable);
  locked.pick_id = `pk_${idHash.slice(0, 20)}`;
  locked.payload_sha256 = await sha256({ ...locked, payload_sha256: undefined });
  return locked;
}

function score(record: UserPick, result: { home_goals: number; away_goals: number }) {
  const actual = outcome(result.home_goals, result.away_goals);
  const points = (scorePick: { home_goals: number; away_goals: number } | null, call: string | null) => {
    const exact = Number(
      scorePick?.home_goals === result.home_goals && scorePick?.away_goals === result.away_goals,
    ) * 3;
    const outcomePoints = Number(call === actual);
    return { exact, outcome: outcomePoints, total: exact + outcomePoints };
  };
  const userBase = points(record.user_pick, record.user_pick.outcome);
  const rivals = record.rivals.map((rival) => ({
    family: rival.family,
    ...points(rival.score_pick, rival.outcome_pick),
  }));
  const active = record.rivals
    .map((rival, index) => (rival.capability === "abstained" ? null : rivals[index].total))
    .filter((value): value is number => value !== null);
  const best = Math.max(0, ...active);
  const beat = active.length > 0 && userBase.total > best;
  return {
    user: { ...userBase, bonus: Number(beat), total: userBase.total + Number(beat) },
    rivals,
    beat_ai: beat,
    best_rival_total: best,
  };
}

async function toView(record: UserPick): Promise<PickView> {
  const all = await matches();
  const match = all.find((row) => row.match_id === record.match.match_id);
  let current = record;
  if (record.status === "draft" && Date.now() >= Date.parse(record.lock_at_utc)) {
    current = await lock(record);
    const store = readStore();
    store[record.match.match_id] = current;
    writeStore(store);
  }
  if (!match && current.status === "locked") {
    return { schema_version: PICK_SCHEMA_VERSION, status: "void", record: current, result: null, scoring: null, preview: true };
  }
  if (match?.is_complete && current.status === "locked") {
    const result = {
      home_goals: match.home_score as number,
      away_goals: match.away_score as number,
      outcome: outcome(match.home_score as number, match.away_score as number),
    };
    return {
      schema_version: PICK_SCHEMA_VERSION,
      status: "scored",
      record: current,
      result,
      scoring: score(current, result),
      preview: true,
    };
  }
  return {
    schema_version: PICK_SCHEMA_VERSION,
    status: current.status,
    record: current,
    result: null,
    scoring: null,
    preview: true,
  };
}

function response(matchId: string, pick: PickView | null, match: MatchRow): PickResponse {
  const lockAt = pick?.record.lock_at_utc ?? match.kickoff_utc;
  return {
    schema_version: PICK_SCHEMA_VERSION,
    match_id: matchId,
    pick,
    editable: !match.is_complete && Date.now() < Date.parse(lockAt) && (!pick || pick.status === "draft"),
    lock_at_utc: lockAt,
    now_utc: new Date(Date.now()).toISOString(),
  };
}

export async function mockFetchPick(matchId: string): Promise<PickResponse | null> {
  const match = (await matches()).find((row) => row.match_id === matchId);
  if (!match) return null;
  const record = readStore()[matchId];
  return response(matchId, record ? await toView(record) : null, match);
}

export async function mockSavePick(
  matchId: string,
  homeGoals: number,
  awayGoals: number,
): Promise<PickResponse> {
  const match = (await matches()).find((row) => row.match_id === matchId);
  if (!match) throw new MockPickError(404, "match_not_found", "Match not found");
  if (match.is_complete) throw new MockPickError(422, "fixture_complete", "This match is final");
  if (!Number.isInteger(homeGoals) || !Number.isInteger(awayGoals) || homeGoals < 0 || awayGoals < 0 || homeGoals > 20 || awayGoals > 20)
    throw new MockPickError(422, "invalid_score", "Scores must be whole numbers from 0 to 20");
  const store = readStore();
  const existing = store[matchId];
  if (existing?.status === "locked" || Date.now() >= Date.parse(existing?.lock_at_utc ?? match.kickoff_utc))
    throw new MockPickError(409, "pick_locked", "This pick locked at kickoff");
  const now = new Date(Date.now()).toISOString();
  const record: UserPick = {
    schema_version: PICK_SCHEMA_VERSION,
    pick_id: null,
    status: "draft",
    match: existing?.match ?? {
      match_id: matchId,
      kickoff_utc: match.kickoff_utc,
      kickoff_time_known: !(match.source_kind === "international" && match.kickoff_utc.endsWith("T00:00:00Z")),
      home_team: match.home_team,
      away_team: match.away_team,
      home_norm: match.home_team.toLocaleLowerCase(),
      away_norm: match.away_team.toLocaleLowerCase(),
      competition: match.competition,
    },
    user_pick: { home_goals: homeGoals, away_goals: awayGoals, outcome: outcome(homeGoals, awayGoals) },
    rivals: MOCK_RIVALS,
    analysis_fingerprint: {
      index_fingerprint: "practice-mode",
      analysis_schema_version: "practice-mode",
      information_cutoff_utc: new Date(Date.parse(match.kickoff_utc) - 1000).toISOString(),
    },
    created_at_utc: existing?.created_at_utc ?? now,
    updated_at_utc: now,
    lock_at_utc: existing?.lock_at_utc ?? match.kickoff_utc,
    locked_at_utc: null,
    payload_sha256: null,
  };
  store[matchId] = record;
  writeStore(store);
  return response(matchId, await toView(record), match);
}

export async function mockDeletePick(matchId: string): Promise<PickResponse | null> {
  const match = (await matches()).find((row) => row.match_id === matchId);
  if (!match) return null;
  const store = readStore();
  const existing = store[matchId];
  if (existing && (existing.status === "locked" || Date.now() >= Date.parse(existing.lock_at_utc)))
    throw new MockPickError(409, "pick_locked", "This pick locked at kickoff");
  delete store[matchId];
  writeStore(store);
  return response(matchId, null, match);
}

export async function mockFetchPicks(limit = 500, offset = 0): Promise<PicksListResponse> {
  const views = await Promise.all(Object.values(readStore()).map(toView));
  views.sort((a, b) => b.record.match.kickoff_utc.localeCompare(a.record.match.kickoff_utc));
  return {
    schema_version: PICK_SCHEMA_VERSION,
    items: views.slice(offset, offset + limit),
    total: views.length,
    limit,
    offset,
  };
}

export async function mockFetchPicksSummary(season: string | null = null): Promise<PicksSummary> {
  const all = (await mockFetchPicks()).items
    .filter((view) => !season || seasonFor(view.record.match.kickoff_utc) === season)
    .sort((a, b) => a.record.match.kickoff_utc.localeCompare(b.record.match.kickoff_utc));
  const counts = { draft: 0, locked: 0, scored: 0, void: 0 };
  const user = { total: 0, exact: 0, outcome: 0, bonus: 0 };
  const rivals = new Map<string, { family: RivalPick["family"]; total: number; exact: number; outcome: number }>();
  const series: PicksSummary["series"] = [];
  let current = 0;
  let best = 0;
  let exactCalls = 0;
  let winnerCalls = 0;
  let mae = 0;
  for (const view of all) {
    counts[view.status] += 1;
    if (view.status !== "scored" || !view.scoring || !view.result) continue;
    for (const key of ["total", "exact", "outcome", "bonus"] as const)
      user[key] += view.scoring.user[key];
    exactCalls += Number(view.scoring.user.exact > 0);
    winnerCalls += Number(view.scoring.user.outcome > 0);
    current = view.scoring.user.outcome > 0 ? current + 1 : 0;
    best = Math.max(best, current);
    mae += Math.abs(
      (view.record.user_pick.home_goals - view.record.user_pick.away_goals) -
        (view.result.home_goals - view.result.away_goals),
    );
    for (const row of view.scoring.rivals) {
      const total = rivals.get(row.family) ?? { family: row.family, total: 0, exact: 0, outcome: 0 };
      total.total += row.total;
      total.exact += row.exact;
      total.outcome += row.outcome;
      rivals.set(row.family, total);
    }
    series.push({
      kickoff_utc: view.record.match.kickoff_utc,
      match_id: view.record.match.match_id,
      user_total: user.total,
      per_family_totals: Object.fromEntries([...rivals].map(([family, row]) => [family, row.total])),
    });
  }
  const scored = counts.scored;
  return {
    schema_version: PICK_SCHEMA_VERSION,
    season,
    counts,
    user,
    rivals: [...rivals.values()],
    series,
    accuracy: { exact: scored ? exactCalls / scored : 0, winner: scored ? winnerCalls / scored : 0 },
    streak: { current, best },
    goal_diff_mae: scored ? mae / scored : 0,
  };
}

function seasonFor(kickoff: string): string {
  const date = new Date(kickoff);
  const start = date.getUTCMonth() >= 6 ? date.getUTCFullYear() : date.getUTCFullYear() - 1;
  return `${start}-${String(start + 1).slice(-2)}`;
}

File: core/golavo_core/picks.py
"""Deterministic fantasy-pick sealing, rival derivation, and scoring.

This module is deliberately pure.  It knows how to turn a leak-safe match
analysis into five rival calls, content-address a locked user pick, score a
finished match, and aggregate season views.  Ledger I/O and kickoff state live
in ``golavo_server.picks``.
"""

from __future__ import annotations

import copy
import hashlib
import json
from collections.abc import Iterable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any, Literal

from jsonschema import Draft202012Validator, FormatChecker

PICK_SCHEMA_VERSION = "0.1.0"
Outcome = Literal["home", "draw", "away"]
OUTCOMES: tuple[Outcome, ...] = ("home", "draw", "away")


def canonical_pick_bytes(record: dict[str, Any]) -> bytes:
    """Return stable UTF-8 JSON bytes without forecast-specific rounding."""
    return json.dumps(
        record,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    ).encode("utf-8")


def pick_id(payload: dict[str, Any]) -> str:
    """Return the content address for a stable locked-pick payload."""
    digest = hashlib.sha256(canonical_pick_bytes(payload)).hexdigest()
    return f"pk_{digest[:20]}"


def pick_payload_sha256(record: dict[str, Any]) -> str:
    """Hash a locked record without its self-referential digest."""
    payload = copy.deepcopy(record)
    payload.pop("payload_sha256", None)
    return hashlib.sha256(canonical_pick_bytes(payload)).hexdigest()


def _stable_pick_payload(record: dict[str, Any]) -> dict[str, Any]:
    payload = copy.deepcopy(record)
    payload.pop("pick_id", None)
    payload.pop("payload_sha256", None)
    return payload


def validate_user_pick(record: dict[str, Any], schema_path: Path | None = None) -> None:
    if schema_path is None:
        from golavo_core.resources import user_pick_schema_path

        schema_path = user_pick_schema_path()
    schema = json.loads(Path(schema_path).read_text(encoding="utf-8"))
    validator = Draft202012Validator(schema, format_checker=FormatChecker())
    validator.evolve(schema=schema["$defs"]["UserPick"]).validate(record)


def verify_pick_integrity(
    record: dict[str, Any], *, expected_id: str | None = None
) -> dict[str, Any]:
    """Validate a locked pick and prove its id, digest, and filename agree."""
    validate_user_pick(record)
    if record.get("status") != "locked":
        raise ValueError("only locked picks carry an immutable integrity stamp")

    stored_hash = record.get("payload_sha256")
    recomputed_hash = pick_payload_sha256(record)
    if stored_hash != recomputed_hash:
        raise ValueError(
            f"payload_sha256 mismatch: stored {stored_hash!r}, content hashes to {recomputed_hash}"
        )

    recomputed_id = pick_id(_stable_pick_payload(record))
    if record.get("pick_id") != recomputed_id:
        raise ValueError(
            f"pick_id mismatch: stored {record.get('pick_id')!r}, content addresses "
            f"to {recomputed_id}"
        )
    if expected_id is not None and expected_id != recomputed_id:
        raise ValueError(
            f"pick filename {expected_id!r} does not match its content id {recomputed_id}"
        )
    return record


def outcome_of(home_goals: int, away_goals: int) -> Outcome:
    if home_goals > away_goals:
        return "home"
    if home_goals < away_goals:
        return "away"
    return "draw"


def _modal_outcome(probs: dict[str, Any]) -> Outcome:
    # Tuple order is the product decision's deterministic home > draw > away
    # tie-break; max returns the first equal value.
    return max(OUTCOMES, key=lambda outcome: float(probs[outcome]))


def derive_rival_picks(analysis: dict[str, Any]) -> dict[str, Any]:
    """Snapshot each council model's honest scoring capability and call."""
    rivals: list[dict[str, Any]] = []
    for model in analysis.get("models", []):
        family = str(model["family"])
        probs = model.get("probs")
        matrix = model.get("score_matrix")
        abstained = bool(model.get("abstained")) or probs is None

        score: dict[str, int] | None = None
        outcome: Outcome | None = None
        capability: Literal["score", "outcome_only", "abstained"]
        if abstained:
            capability = "abstained"
        elif matrix is not None:
            likely = matrix["most_likely"]
            # Stored ScoreMatrix uses ``home``/``away``. Accept the explicit
            # ``*_goals`` spelling as well for contract fixtures and migrations.
            home = int(likely.get("home", likely.get("home_goals")))
            away = int(likely.get("away", likely.get("away_goals")))
            score = {"home_goals": home, "away_goals": away}
            outcome = outcome_of(home, away)
            capability = "score"
        else:
            outcome = _modal_outcome(probs)
            capability = "outcome_only"

        rivals.append(
            {
                "family": family,
                "capability": capability,
                "score_pick": score,
                "outcome_pick": outcome,
            }
        )

    return {
        "analysis_fingerprint": {
            "index_fingerprint": str(
                analysis.get("index_fingerprint") or analysis.get("_index_fingerprint") or "unknown"
            ),
            "analysis_schema_version": str(analysis.get("schema_version") or "unknown"),
            "information_cutoff_utc": str(
                analysis.get("information_cutoff_utc") or "1970-01-01T00:00:00Z"
            ),
        },
        "rivals": rivals,
    }


def _points_for_call(
    *,
    score: dict[str, Any] | None,
    outcome: str | None,
    result_home: int,
    result_away: int,
) -> dict[str, int]:
    actual = outcome_of(result_home, result_away)
    exact = (
        int(
            score is not None
            and int(score["home_goals"]) == result_home
            and int(score["away_goals"]) == result_away
        )
        * 3
    )
    outcome_points = int(outcome == actual)
    return {"exact": exact, "outcome": outcome_points, "total": exact + outcome_points}


def score_pick(locked_record: dict[str, Any], result: dict[str, Any]) -> dict[str, Any]:
    """Score the user and all frozen rivals under the same additive rules."""
    result_home = int(result["home_goals"])
    result_away = int(result["away_goals"])
    user_call = locked_record["user_pick"]
    user_base = _points_for_call(
        score={
            "home_goals": int(user_call["home_goals"]),
            "away_goals": int(user_call["away_goals"]),
        },
        outcome=user_call.get("outcome")
        or outcome_of(int(user_call["home_goals"]), int(user_call["away_goals"])),
        result_home=result_home,
        result_away=result_away,
    )

    rival_points: list[dict[str, Any]] = []
    played_totals: list[int] = []
    for rival in locked_record.get("rivals", []):
        points = _points_for_call(
            score=rival.get("score_pick"),
            outcome=rival.get("outcome_pick"),
            result_home=result_home,
            result_away=result_away,
        )
        if rival.get("capability") != "abstained":
            played_totals.append(points["total"])
        rival_points.append({"family": rival["family"], **points})

    best_rival = max(played_totals, default=0)
    beat_ai = bool(played_totals) and user_base["total"] > best_rival
    bonus = int(beat_ai)
    return {
        "user": {
            "exact": user_base["exact"],
            "outcome": user_base["outcome"],
            "bonus": bonus,
            "total": user_base["total"] + bonus,
        },
        "rivals": rival_points,
        "beat_ai": beat_ai,
        "best_rival_total": best_rival,
    }


def football_season(kickoff_utc: str) -> str:
    kickoff = datetime.fromisoformat(kickoff_utc.replace("Z", "+00:00"))
    kickoff = kickoff.astimezone(UTC)
    first_year = kickoff.year if (kickoff.month, kickoff.day) >= (7, 1) else kickoff.year - 1
    return f"{first_year}-{str(first_year + 1)[-2:]}"


def _ordered_views(views: Iterable[dict[str, Any]]) -> list[dict[str, Any]]:
    return sorted(
        views,
        key=lambda view: (
            view["record"]["match"]["kickoff_utc"],
            view["record"]["match"]["match_id"],
        ),
    )


def season_summary(views: list[dict[str, Any]], *, season: str | None = None) -> dict[str, Any]:
    """Aggregate ledger views without persisting derived standings."""
    filtered = [
        view
        for view in views
        if season is None or football_season(view["record"]["match"]["kickoff_utc"]) == season
    ]
    ordered = _ordered_views(filtered)
    counts = {status: 0 for status in ("draft", "locked", "scored", "void")}
    for view in ordered:
        status = str(view["status"])
        if status in counts:
            counts[status] += 1

    scored = [view for view in ordered if view.get("status") == "scored"]
    user = {"total": 0, "exact": 0, "outcome": 0, "bonus": 0}
    rival_totals: dict[str, dict[str, Any]] = {}
    series: list[dict[str, Any]] = []
    goal_diff_errors: list[int] = []
    best_streak = 0
    current_streak = 0

    for view in scored:
        scoring = view["scoring"]
        for key in user:
            user[key] += int(scoring["user"][key])
        for rival in scoring.get("rivals", []):
            row = rival_totals.setdefault(
                str(rival["family"]),
                {"family": str(rival["family"]), "total": 0, "exact": 0, "outcome": 0},
            )
            for key in ("total", "exact", "outcome"):
                row[key] += int(rival[key])

        if int(scoring["user"]["outcome"]) >= 1:
            current_streak += 1
            best_streak = max(best_streak, current_streak)
        else:
            current_streak = 0

        call = view["record"]["user_pick"]
        result = view["result"]
        goal_diff_errors.append(
            abs(
                (int(call["home_goals"]) - int(call["away_goals"]))
                - (int(result["home_goals"]) - int(result["away_goals"]))
            )
        )
        series.append(
            {
                "kickoff_utc": view["record"]["match"]["kickoff_utc"],
                "match_id": view["record"]["match"]["match_id"],
                "user_total": user["total"],
                "per_family_totals": {
                    family: row["total"] for family, row in sorted(rival_totals.items())
                },
            }
        )

    scored_count = len(scored)
    exact_calls = sum(int(view["scoring"]["user"]["exact"] > 0) for view in scored)
    winner_calls = sum(int(view["scoring"]["user"]["outcome"] > 0) for view in scored)
    return {
        "schema_version": PICK_SCHEMA_VERSION,
        "season": season,
        "counts": counts,
        "user": user,
        "rivals": [rival_totals[key] for key in sorted(rival_totals)],
        "series": series,
        "accuracy": {
            "exact": exact_calls / scored_count if scored_count else 0.0,
            "winner": winner_calls / scored_count if scored_count else 0.0,
        },
        "streak": {"current": current_streak, "best": best_streak},
        "goal_diff_mae": (
            sum(goal_diff_errors) / len(goal_diff_errors) if goal_diff_errors else 0.0
        ),
    }
